odxtools/cli/decode.py

```diff
@@ -0,0 +1,130 @@
+    else:
+        return str(v)
+
+
+def print_summary(
+    odxdb: Database,
+    ecu_variants: Optional[List[str]] = None,
+    data: bytes = b'',
+    decode: bool = False,
+) -> None:
+    ecu_names = ecu_variants if ecu_variants else [ecu.short_name for ecu in odxdb.ecus]
+    services: Dict[DiagService, List[str]] = {}
+    for ecu_name in ecu_names:
+        ecu = odxdb.ecus[ecu_name]
+        if not ecu:
+            print(f"The ecu variant '{ecu_name}' could not be found!")
+            continue
+        if data:
+            found_services = ecu._find_services_for_uds(data)
+            for found_service in found_services:
+                ecu_names = services.get(found_service, [])
+                ecu_names.append(ecu_name)
+                services[found_service] = ecu_names
+
+    print(f"Binary data: {data.hex(' ')}")
+    for service, ecu_names in services.items():
+        if isinstance(service, DiagService):
+            print(
+                f"Decoded by service '{service.short_name}' (decoding ECUs: {', '.join(ecu_names)})"
+            )
+        elif isinstance(service, SingleEcuJob):
+            print(
+                f"Decoded by single ecu job '{service.short_name}' (decoding ECUs: {', '.join(ecu_names)})"
+            )
+        else:
+            print(f"Decoded by unknown diagnostic communication: '{service.short_name}' "
+                  f"(decoding ECUs: {', '.join(ecu_names)})")
+
+        if decode:
+            if data is None:
+                odxraise("Data is required for decoding")
+
+            decoded = service.decode_message(data)
+            print(f"Decoded data:")
+            for param_name, param_value in decoded.param_dict.items():
+                print(f"  {param_name}={get_display_value(param_value)}")
+
+
+def add_subparser(subparsers: "argparse._SubParsersAction") -> None:
+    parser = subparsers.add_parser(
+        "decode",
+        description="\n".join([
+            "Decode request by hex-data",
+            "",
+            "Examples:",
+            "  For displaying the service associated with the request 10 01 & decoding it:",
+            "    odxtools decode ./path/to/database.pdx -D -d 10 01",
+            "  For displaying the service associated with the request 10 01, without decoding it:",
+            "    odxtools decode ./path/to/database.pdx -d 10 01",
+            "  For more information use:",
+            "    odxtools decode -h",
+        ]),
+        help="Find & print service by hex-data. Can also decode the hex-data to named parameters.",
+        formatter_class=argparse.RawTextHelpFormatter,
+    )
+    _parser_utils.add_pdx_argument(parser)
+
+    parser.add_argument(
+        "-v",
+        "--variants",
+        nargs=1,
+        metavar="VARIANT",
+        required=False,
+        help="Specifies which ecu variants should be included.",
+        default="all",
+    )
+
+    parser.add_argument(
+        "-d",
+        "--data",
+        nargs="*",
```

Comment:
I'd vastly prefer to make this a simple string, because right now specifying positional arguments after -D makes things quite iffy (e.g. `odxtools decode -D 01 01 my_file.pdx` will fail while `odxtools decode -D "01 01" my_file.pdx` will work just fine).
```suggestion
```

Reply:
changed it